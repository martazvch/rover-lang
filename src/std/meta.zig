const Value = @import("../runtime/values.zig").Value;
const Type = @import("../frontend/type_system.zig").Type;

pub const NativeFn = *const fn ([]const Value) Value;

pub const NativeFnMeta = struct {
    params: []const Type = &.{},
    return_type: Type,
    function: NativeFn,
};

pub const ModuleMeta = struct {
    name: []const u8,
};
