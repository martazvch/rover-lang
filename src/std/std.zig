pub const testing = @import("testing.zig");
pub const time = @import("time/time.zig");
