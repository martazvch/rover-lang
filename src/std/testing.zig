pub const assert = @import("testing/testing.zig").assert;
pub const eql_int = @import("testing/testing.zig").eql_int;
