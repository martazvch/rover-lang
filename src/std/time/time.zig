const std = @import("std");
const Value = @import("../../runtime/values.zig").Value;
const Meta = @import("../meta.zig");
const NativeFn = Meta.NativeFn;
const NativeFnMeta = Meta.NativeFnMeta;
const ModuleMeta = Meta.ModuleMeta;

pub const meta = ModuleMeta{
    .name = "time",
};

// Clock
pub const clock = NativeFnMeta{
    .params = &.{},
    .return_type = .float,
    .function = _clock,
};

pub fn _clock(_: []const Value) Value {
    return .{ .float = @as(f64, @floatFromInt(std.time.milliTimestamp())) / 1000.0 };
}
