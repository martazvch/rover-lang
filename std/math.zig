pub fn cos(x: f64) f64 {
    return @cos(x);
}

pub fn sin(x: f64) f64 {
    return @sin(x);
}
