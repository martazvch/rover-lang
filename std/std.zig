pub const time = @import("time.zig");
pub const math = @import("math.zig");
