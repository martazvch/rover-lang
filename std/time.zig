const std = @import("std");

pub fn _clock() f64 {
    return @as(f64, @floatFromInt(std.time.milliTimestamp())) / 1000.0;
}
