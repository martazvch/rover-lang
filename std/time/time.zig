const std = @import("std");
const Value = @import("../../src/runtime/values.zig").Value;
const TypeSys = @import("../../src/frontend/type_system.zig");
const NativeFnMeta = @import("../meta.zig").NativeFnMeta;

// Clock
pub const NativeClock = NativeFnMeta{
    .params = .{},
    .return_type = TypeSys.Float,
    .function = clock_native,
};

pub fn clock_native(_: []const Value) Value {
    return .{ .Float = @as(f64, @floatFromInt(std.time.milliTimestamp())) / 1000.0 };
}

// Test
pub const NativeTest = NativeFnMeta{
    .params = .{ TypeSys.Int, TypeSys.Float },
    .return_type = TypeSys.Bool,
    .function = test_native,
};

pub fn test_native(values: []const Value) Value {
    std.debug.print(values[0].Int.print);
    std.debug.print(values[1].Float.print);

    return Value.bool_(true);
}
