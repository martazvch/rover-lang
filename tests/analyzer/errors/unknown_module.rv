use std.blender
use ctd.blender

fn main() {
    {
        use std.time
    }

    print clock()
}