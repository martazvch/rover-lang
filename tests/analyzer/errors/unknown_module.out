// IR from the valid import
[Structure declaration Vec2, index: 0, scope: global]
[Fn declaration introduce, index: 1, scope: global, return kind: implicit_void]
    [Print]
        [String This is a mathematical module]
[Fn declaration getVecFromPos, index: 2, scope: global, return kind: implicit_value]
    [Structure literal]
        - structure
        [Variable index: 0, scope: global]
        - args
        [Variable index: 1, scope: local]
        [Variable index: 2, scope: local]

missing_file_in_cwd, pipou.rv
missing_file_in_module, mathh.rv, data
undeclared_var, math
